### ISSUE TITLE
useFragment/useFragment_experimental - add `skip` option

### ISSUE THREAD

**Issue by tvler** [apollographql/apollo-feature-requests#350]
- Add a skip option that behaves similarly to the skip option on useQuery
- This is useful when you want to call useFragment in a component, but only if an ID is passed down by a parent (similar to useQuery)

**Comment by pfcodes** [comment]
+1, much needed

**Comment by nathggns** [comment]
+1!

**Comment by nathggns** [comment]
I'm seeing useFragment log errors in my tests (vitest) when the from reference doesn't include fields required to generate a cache ID – but the value I'm using as the primary key may be undefined within my component. In theory I could restructure my component structure so the hook doesn't run if the the value isn't defined, but in reality useFragment is being called deep within multiple levels of hooks, so it's actually quite difficult / tedious to do that. 

```
Invariant Violation: Missing field 'number' while extracting keyFields from {
  "__typename": "[REDACTED]"
  "number": <undefined>
}
```

**Comment by phryneas** [comment]
Hmm, we could certainly add something more ergonomic there - but as for now, I believe you should be able to pass a string (e.g. "__skip_nonexistent") into `from` if you don't have an object for identification.

**Comment by nathggns** [comment]
Yes, this is what we've done essentially, but would be good for a native way to do this.

**Comment by SimpleCreations** [comment]
Reviving this thread! There's no straightforward way to do this currently, as we can do with `useQuery`.

Forced to do an ugly workaround like this:
```ts
const { data, complete } = useFragment<
    CategoryFieldsFragment,
    { countryCode: CountryCode }
  >({
    fragment: CategoryFieldsFragmentDoc,
    ...(countryCode != null
      ? {
          from: { __typename: 'Category', id },
          variables: { countryCode },
        }
      : { from: '__skip' }),
  });
```

Here's how it would look with a `skip` option:
```ts
const { data, complete } = useFragment<
    CategoryFieldsFragment,
    { countryCode: CountryCode }
  >(
    countryCode != null
      ? {
          fragment: CategoryFieldsFragmentDoc,
          from: { __typename: 'Category', id },
          variables: { countryCode },
        }
      : { skip: true },
  );
```

**Comment by phryneas** [comment]
Currently, you can just pass `from: null` in (even without a fragment, although TS might complain about that), so you could do this:
```ts

const skipToken = { from: null } as any as UseFragmentOptions


const { data, complete } = useFragment<
    CategoryFieldsFragment,
    { countryCode: CountryCode }
  >(
    countryCode != null
      ? {
          fragment: CategoryFieldsFragmentDoc,
          from: { __typename: 'Category', id },
          variables: { countryCode },
        }
      : skipToken,
  );
```

**Comment by SimpleCreations** [comment]
Thanks @phryneas! `from: null` works, but the TypeScript types require the `fragment` property to be present... So, some sort of workaround is still required — either a `...` spread like in my previous comment, or some sort of type suppression with `any`/`as` etc...

**Comment by phryneas** [comment]
Since `fragment` will never be read if you pass in `from: null`, you can use any arbitrary fragment, or even an empty document.

This should work without any type assertions, and you'll have to define that `skipToken` only once in your whole application:
```ts
    const skipToken = {
      fragment: {
        kind: Kind.DOCUMENT,
        definitions: [],
      } satisfies DocumentNode,
      from: null,
    };
```
